llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDKernelAttrs.cpp

```diff
@@ -0,0 +1,39 @@
+//===---- LowerESIMDKernelAttrs - lower __esimd_set_kernel_attributes ---===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Finds and adds  sycl_explicit_simd attributes to wrapper functions that wrap
+// ESIMD kernel functions
+
+#include "llvm/SYCLLowerIR/ESIMD/ESIMDUtils.h"
+#include "llvm/SYCLLowerIR/ESIMD/LowerESIMD.h"
+
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+
+#define DEBUG_TYPE "LowerESIMDKernelAttrs"
+
+using namespace llvm;
+
+namespace llvm {
+PreservedAnalyses
+SYCLLowerESIMDKernelAttrPass::run(Module &M, ModuleAnalysisManager &MAM) {
+  bool Modified = false;
+  for (Function &F : M) {
+    if (llvm::esimd::isESIMD(F)) {
+      llvm::esimd::traverseCallgraphUp(&F, [&](Function *GraphNode) {
```

Comment:
BTW, is the wrapper function a `spir_kernel` or `spir_function`? Is the user function  a `spir_kernel` or `spir_function` in presence of wrapper?

Reply:
For the record: as we talked, the user function is `spir_func`, which is called by two kernels with `spir_kernel` CC - the normal and the one with the rounded range.